src/machine/machine_mimxrt1062_i2c.go

```diff
@@ -0,0 +1,725 @@
+	}
+	// if we end up with zero cycles, then set the filter to a single cycle unless
+	// the bus clock is greater than 10x the desired glitch width
+	if (cycles == 0) && (busCycles <= (width * 10)) {
+		cycles = 1
+	}
+	return cycles
+}
+
+// checkStatus converts the status register to a resultFlag for return, and
+// clears any errors if present.
+func (i2c *I2C) checkStatus(status statusFlag) resultFlag {
+	result := resultSuccess
+	// check for error. these errors cause a stop to be sent automatically.
+	// we must clear the errors before a new transfer can start.
+	if status &= statusError; 0 != status {
+		// select the correct error code ordered by severity, bus issues first.
+		if 0 != (status & statusPinLowTimeout) {
+			result = resultPinLowTimeout
+		} else if 0 != (status & statusArbitrationLost) {
+			result = resultArbitrationLost
+		} else if 0 != (status & statusNackDetect) {
+			result = resultNak
+		} else if 0 != (status & statusFifoErr) {
+			result = resultFifoError
+		}
+		// clear the flags
+		i2c.Bus.MSR.Set(uint32(status))
+		// reset fifos. these flags clear automatically.
+		i2c.Bus.MCR.SetBits(nxp.LPI2C_MCR_RRF | nxp.LPI2C_MCR_RTF)
+	}
+	return result
+}
+
+func (i2c *I2C) getFIFOSize() (rx, tx uint32) { return 4, 4 }
+func (i2c *I2C) getFIFOCount() (rx, tx uint32) {
+	mfsr := i2c.Bus.MFSR.Get()
+	return (mfsr & nxp.LPI2C_MFSR_RXCOUNT_Msk) >> nxp.LPI2C_MFSR_RXCOUNT_Pos,
+		(mfsr & nxp.LPI2C_MFSR_TXCOUNT_Msk) >> nxp.LPI2C_MFSR_TXCOUNT_Pos
+}
+
+func (i2c *I2C) waitForTxReady() resultFlag {
+	result := resultSuccess
+	_, txSize := i2c.getFIFOSize()
+	for {
+		_, txCount := i2c.getFIFOCount()
+		status := statusFlag(i2c.Bus.MSR.Get())
+		if result = i2c.checkStatus(status); resultSuccess != result {
+			break
+		}
+		if txSize-txCount > 0 {
+			break
+		}
+	}
+	return result
+}
+
+func (i2c *I2C) waitForTxEmpty() resultFlag {
+	result := resultSuccess
+	for {
+		_, txCount := i2c.getFIFOCount()
+		status := statusFlag(i2c.Bus.MSR.Get())
+		if result = i2c.checkStatus(status); resultSuccess != result {
+			break
+		}
+		if 0 == txCount {
+			break
+		}
+	}
+	return result
+}
+
+// isBusBusy checks if the I2C bus is busy, returning true if it is busy and we
+// are not the ones driving it, otherwise false.
+func (i2c *I2C) isBusBusy() bool {
+	status := statusFlag(i2c.Bus.MSR.Get())
+	return (0 != (status & statusBusBusy)) && (0 == (status & statusBusy))
+}
+
+// addressable represents a type that can provide fully-formatted I2C peripheral
+// addresses for both read operations and write operations.
+type addressable interface {
+	toRead() uint32
+	toWrite() uint32
+	bitSize() uint8
+	shift(I2CDirection) uint32
+}
+
+// address7Bit and address10Bit stores the unshifted original I2C peripheral
+// address in an unsigned integral data type and implements the addressable
+// interface to reformat addresses as required for read/write operations.
+//   TODO:
+//    add 10-bit address support
+type (
+	address7Bit uint8
```

Comment:
It looks like you did this for future compatibility?

I am generally hesitant to include code that does not currently serve a purpose, especially when it complicates things (here with an interface value, even). Right now, only 7 bit addresses are supported.

Take a look at this code from the AVR:

https://github.com/tinygo-org/tinygo/blob/3cdc110462369500aedfc1511f7c409c49200e35/src/machine/machine_atmega.go#L70-L75

To me, this is much simpler and much easier to understand. It might need revisiting once 10 bit addresses get supported (although I haven't seen any such devices yet) but those 3 LOC for address calculation are all that's really needed.

Reply:
Also have never encountered a device with 10-bit address. Will refactor to remove this address generalization

Reply:
Not sure if it was intentional, but there is an inconsistency with the type of peripheral addresses in the `machine` package.

In `machine/i2c.go`, we have the general purpose `ReadRegister` and `WriteRegister` that would both be unable to support future 10-bit addresses (which their godoc comments rightfully warn the user) with a `uint8` parameter:

```go
func (i2c I2C) WriteRegister(address uint8, register uint8, data []byte) error
func (i2c I2C) ReadRegister(address uint8, register uint8, data []byte) error
```

Yet all of the machine implementation-specific `Tx` methods (which the above depend on) provision future 10-bit support with a `uint16` parameter:

```
$ grep -RF 'I2C) Tx('
 src/machine/machine_k210.go:func (i2c I2C) Tx(addr uint16, w, r []byte) error {
 src/machine/machine_mimxrt1062_i2c.go:func (i2c I2C) Tx(addr uint16, w, r []byte) error {
 src/machine/machine_attiny.go:func (i2c I2C) Tx(addr uint16, w, r []byte) error {
 src/machine/machine_atsamd21.go:func (i2c I2C) Tx(addr uint16, w, r []byte) error {
 src/machine/machine_atmega.go:func (i2c I2C) Tx(addr uint16, w, r []byte) error {
 src/machine/machine_generic.go:func (i2c I2C) Tx(addr uint16, w, r []byte) error {
 src/machine/machine_stm32_i2c.go:func (i2c I2C) Tx(addr uint16, w, r []byte) error {
 src/machine/machine_atsamd51.go:func (i2c I2C) Tx(addr uint16, w, r []byte) error {
 src/machine/machine_stm32f103xx.go:func (i2c I2C) Tx(addr uint16, w, r []byte) error {
 src/machine/machine_fe310.go:func (i2c I2C) Tx(addr uint16, w, r []byte) error {
 src/machine/machine_nrf.go:func (i2c I2C) Tx(addr uint16, w, r []byte) (err error) {
```

This seems backwards to me. Specifically, the _machine-independent_ API `ReadRegister`/`WriteRegister` should be insulated from future changes that add 10-bit addressing support by simply widening their `address` type to `uint16`.

The 7-bit-only policy should instead be considered a defect (or at least an incomplete implementation), enforced by convention and documentation, so that adding 10-bit support in the future is a transparent upgrade to current users and would not become an API-breaking change.

Reply:
Well, yeah you are right in that it is inconsistent. However, I do not agree that one or the other is necessarily a defect, they are just inconsistent with each other.

It might be worth changing the API here, or even removing them entirely as `WriteRegister` forces a heap allocation which you generally wouldn't want (even though these methods are very convenient). In any case, this discussion is somewhat separate from this PR.

This might be an interesting read: https://www.kernel.org/doc/Documentation/i2c/ten-bit-addresses

This as well: https://github.com/rust-embedded/embedded-hal/issues/147